js/imageeditor/shared/tools/Sources.svelte

```diff
@@ -0,0 +1,194 @@
+
+	let active_mode: "webcam" | "color" | null = null;
+	let background: Blob | File | null;
+
+	const sources_meta = {
+		upload: {
+			icon: UploadIcon,
+			label: "Upload",
+			order: 0,
+			id: "bg_upload",
+			cb() {
+				upload.open_file_upload();
+			}
+		},
+		webcam: {
+			icon: WebcamIcon,
+			label: "Webcam",
+			order: 1,
+			id: "bg_webcam",
+			cb() {
+				active_mode = "webcam";
+			}
+		},
+		clipboard: {
+			icon: ImagePaste,
+			label: "Paste",
+			order: 2,
+			id: "bg_clipboard",
+			cb() {
+				process_clipboard();
+			}
+		}
+	} as const;
+
+	$: sources_list = sources
+		.map((src) => sources_meta[src])
+		.sort((a, b) => a.order - b.order);
+
+	let upload: Upload;
+
+	async function process_clipboard(): Promise<void> {
+		const items = await navigator.clipboard.read();
+
+		for (let i = 0; i < items.length; i++) {
+			const type = items[i].types.find((t) => t.startsWith("image/"));
+			if (type) {
+				const blob = await items[i].getType(type);
+
+				background = blob || null;
+			}
+		}
+	}
+
+	function handle_upload(e: CustomEvent<Blob>): void {
+		const file_data = e.detail;
+		background = file_data;
+		active_mode = null;
+	}
+
+	let should_reset = true;
+
+	async function set_background(): Promise<void> {
+		if (!$pixi) return;
+		if (background) {
+			if (should_reset) {
+				$pixi?.reset?.();
+			}
+
+			const add_image = add_bg_image(
+				$pixi.background_container,
+				$pixi.renderer,
+				background,
+				$pixi.resize
+			);
+
+			$dimensions = await add_image.start();
+			add_image.execute();
+
+			should_reset = true;
+		}
+	}
+
+	async function process_bg_file(file: FileData | null): Promise<void> {
+		if (!file || !file.url) return;
+		should_reset = false;
+
+		const blob_res = await fetch(file.url);
+		const blob = await blob_res.blob();
+		background = blob;
+	}
+
+	function handle_key(e: KeyboardEvent): void {
+		if (e.key === "Escape") {
+			active_mode = null;
+		}
+	}
+
+	$: background && set_background();
+	$: process_bg_file(background_file);
+
+	onMount(() => {
+		tick().then(() => {
+			if (!$pixi) return;
+			function reset(): void {}
+			$pixi = { ...$pixi, reset };
+		});
+
+		return register_tool("bg", {
+			default: "bg_upload",
+			options: sources_list || []
+		});
+	});
+</script>
+
+<svelte:window on:keydown={handle_key} />
+
+{#if $active_tool === "bg"}
+	<div class="upload-container">
+		<Upload
+			hidden={true}
+			bind:this={upload}
+			filetype="image/*"
+			on:load={handle_upload}
+			on:error
+			{root}
+			disable_click={!sources.includes("upload")}
+			format="blob"
+		></Upload>
+		{#if active_mode === "webcam"}
+			<div class="modal">
+				<div class="modal-inner">
+					<Webcam
+						on:capture={handle_upload}
+						on:error
+						on:drag
+						{mirror_webcam}
+						streaming={false}
+						mode="image"
+						include_audio={false}
+						{i18n}
+					/>
+				</div>
+			</div>
+		{/if}
+	</div>
+{/if}
+
+<style>
+	.modal {
+		position: fixed;
```

Comment:
This probably isn't going to work on spaces. We'll need to do something more dynamic there using position absolute and the current scroll position. 

Probably a good time to fix the toasts positions because it is the same problem.